{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "import jupyter_black\n",
    "\n",
    "jupyter_black.load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import sys\n",
    "import time\n",
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "from pyfaidx import Fasta\n",
    "\n",
    "base = Path(\"../../data/Pla2g2/\")\n",
    "raw = base / \"raw\"\n",
    "data_info_excel = raw / \"SM2.xlsx\"\n",
    "fasta_alignment = raw / \"SM9_Pla2g2_alignment.fasta\"\n",
    "\n",
    "fasta_out = base / \"Pla2g2.fasta\"\n",
    "csv_out = base / \"Pla2g2.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- create FASTA file ---\n",
    "fasta_handler = Fasta(fasta_alignment)\n",
    "with open(fasta_out, \"w\") as fasta_handle:\n",
    "    for header, seq in fasta_handler.items():\n",
    "        if header.startswith(\"---\"):\n",
    "            continue\n",
    "        seq = str(seq).replace(\"-\", \"\")\n",
    "        fasta_handle.write(f\">{header}\\n\")\n",
    "        fasta_handle.write(f\"{seq}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel(data_info_excel, sheet_name=\"Dataset Key\")\n",
    "df = df.rename(\n",
    "    columns={\"Name in the dataset\": \"uid\", \"Species\": \"taxa_name\", \"Gene\": \"gene\"}\n",
    ")\n",
    "df = df[[\"uid\", \"gene\", \"strand\", \"taxa_name\"]]\n",
    "df[[\"gene\", \"gene_info\"]] = df[\"gene\"].str.split(\" \", n=1, expand=True)\n",
    "\n",
    "taxa_rename = {\n",
    "    \"Ailurus fulgens styani\": \"Ailurus styani\",\n",
    "    \"Anser cygnoides\": \"Anser cygnoid\",\n",
    "    \"Apteryx australis mantelli\": \"Apteryx mantelli mantelli\",\n",
    "    \"Crotalus helleri\": \"Crotalus oreganus helleri\",\n",
    "    \"Deinakgistrodon acutus\": \"Deinagkistrodon acutus\",\n",
    "    \"Nannopterum auritus\": \"Nannopterum\",  # (species not in UniProt)\n",
    "    \"Ophisaurus gracilis\": \"Dopasia gracilis\",\n",
    "    \"Urile pelagicus\": \"Phalacrocorax\",  # (species not in UniProt) Urile penicillatus\n",
    "}\n",
    "for old_name, new_name in taxa_rename.items():\n",
    "    df.loc[df[\"taxa_name\"] == old_name, \"taxa_name\"] = new_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper function to download data\n",
    "def get_url(url, **kwargs):\n",
    "    response = requests.get(url, **kwargs)\n",
    "\n",
    "    if not response.ok:\n",
    "        print(response.text)\n",
    "        response.raise_for_status()\n",
    "        sys.exit()\n",
    "\n",
    "    return response\n",
    "\n",
    "class TaxonAPI:\n",
    "    def __init__(self, taxon_mapper_path: Path) -> None:\n",
    "        # documentaion: https://www.ebi.ac.uk/proteins/api/doc/\n",
    "        # limit = 200 requests/second/user\n",
    "        self.base_url = \"https://www.ebi.ac.uk/proteins/api/taxonomy\"\n",
    "        # TODO experiment with SQLite\n",
    "        self.taxon_mapper_path = taxon_mapper_path\n",
    "        self.df_taxon_mapper = self._load_taxon_mapper()\n",
    "\n",
    "    def _load_taxon_mapper(self):\n",
    "        # taxon_mapper_file = raw / \"taxon_mapper.csv\"\n",
    "        if not self.taxon_mapper_file.is_file():\n",
    "            df_taxon_mapper = pd.DataFrame()\n",
    "        else:\n",
    "            df_taxon_mapper = pd.read_csv(self.taxon_mapper_file)\n",
    "        return df_taxon_mapper\n",
    "\n",
    "    def _get_taxa_id(self, taxon_name: str) -> tuple[str, str]:\n",
    "        \"\"\"For a given taxa name, search and return its taxa ID.\n",
    "\n",
    "        Return taxa ID (int) and its rank if taxon_name exists, else return None.\n",
    "        \"\"\"\n",
    "        # make request\n",
    "        url = f\"{self.base_url}/name/{taxon_name}\"\n",
    "        response = get_url(url)\n",
    "        try:\n",
    "            taxonomies = response.json()[\"taxonomies\"]\n",
    "        except requests.HTTPError as err:\n",
    "            txt = err\n",
    "            if str(txt).startswith(\"404 Client Error: Not Found for url\"):\n",
    "                taxonomies = None\n",
    "            else:\n",
    "                raise Exception(err)\n",
    "\n",
    "        # get taxa_id\n",
    "        if taxonomies is None:\n",
    "            taxa_id, rank = None, None\n",
    "        elif len(taxonomies) == 1:\n",
    "            taxa_id = int(taxonomies[0][\"taxonomyId\"])\n",
    "            rank = taxonomies[0][\"rank\"]\n",
    "        else:\n",
    "            raise Exception(f\"{taxon_name}: found more than one taxa.\")\n",
    "        return taxa_id, rank\n",
    "\n",
    "    def _get_taxa_rank(self, taxa_id: str, rank) -> str:\n",
    "        url = f\"{self.base_url}/lineage/{taxa_id}\"\n",
    "        response = get_url(url=url)\n",
    "\n",
    "        rank_name = \"None\"\n",
    "        for taxonomy in response.json()[\"taxonomies\"]:\n",
    "            if taxonomy[\"rank\"] == rank:\n",
    "                rank_name = taxonomy[\"scientificName\"]\n",
    "                break\n",
    "        return rank_name\n",
    "        \n",
    "    def fill_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# get species that are in df but not in taxon_mapper\n",
    "if \"taxon_name\" in df_taxon_mapper:\n",
    "    taxon_lst = df_taxon_mapper[\"taxon_name\"].to_list()\n",
    "else:\n",
    "    taxon_lst = []\n",
    "unknown_taxon = df.loc[~df[\"taxa_name\"].isin(taxon_lst), \"taxa_name\"].unique()\n",
    "\n",
    "\n",
    "ranks = [\"class\", \"order\", \"clade\", \"family\", \"genus\", \"species\"]\n",
    "api_call_counter = 0\n",
    "for idx, taxon_name in enumerate(unknown_taxon):\n",
    "    df_idx = len(df_taxon_mapper)\n",
    "    taxa_id, rank = get_taxa_id(taxon_name=taxon_name)\n",
    "    new_row = pd.DataFrame(\n",
    "        {\"taxon_id\": taxa_id, \"taxon_name\": taxon_name, rank: taxon_name},\n",
    "        index=[df_idx],\n",
    "    )\n",
    "    df_taxon_mapper = pd.concat([new_row, df_taxon_mapper.loc[:]])\n",
    "    for rank in ranks:\n",
    "        for idx, row in df_taxon_mapper.iterrows():\n",
    "            if (rank in row) and pd.notna(row[rank]):\n",
    "                continue\n",
    "            api_call_counter += 1\n",
    "            taxa = get_taxa_rank(taxa_id=row[\"taxon_id\"], rank=rank)\n",
    "            df_taxon_mapper.loc[idx, rank] = taxa\n",
    "            if api_call_counter % 200 == 0:\n",
    "                time.sleep(1)\n",
    "    if (idx + 1) % 200 == 0:\n",
    "        time.sleep(1)\n",
    "    break\n",
    "# df_taxon_mapper[\"taxon_id\"] = df_taxon_mapper[\"taxon_id\"].astype(int)\n",
    "\n",
    "\n",
    "# get additional taxonomies\n",
    "ranks = [\"class\", \"order\", \"family\", \"genus\", \"species\"]\n",
    "api_call_counter = 0\n",
    "\n",
    "df_taxon_mapper.to_csv(taxon_mapper_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>species</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>424585</td>\n",
       "      <td>Ailurus styani</td>\n",
       "      <td>Ailurus styani</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Carnivora</td>\n",
       "      <td>Ailuridae</td>\n",
       "      <td>Ailurus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9646</td>\n",
       "      <td>Ailuropoda melanoleuca</td>\n",
       "      <td>Ailuropoda melanoleuca</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Carnivora</td>\n",
       "      <td>Ursidae</td>\n",
       "      <td>Ailuropoda</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   taxon_id              taxon_name                 species     class  \\\n",
       "1    424585          Ailurus styani          Ailurus styani  Mammalia   \n",
       "0      9646  Ailuropoda melanoleuca  Ailuropoda melanoleuca  Mammalia   \n",
       "\n",
       "       order     family       genus  \n",
       "1  Carnivora  Ailuridae     Ailurus  \n",
       "0  Carnivora    Ursidae  Ailuropoda  "
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_taxon_mapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxas = df_taxon_mapper.columns.to_list()[1:]\n",
    "cols = [\"species\", \"genus\", \"family\", \"order\"]\n",
    "for idx, row in df.iterrows():\n",
    "    species = row[\"taxa_name\"]\n",
    "    for taxa in taxas:\n",
    "        if species in df_taxon_mapper[taxa].to_list():\n",
    "            df.loc[idx, cols] = df_taxon_mapper.loc[\n",
    "                df_taxon_mapper[taxa] == species, cols\n",
    "            ].iloc[0]\n",
    "            break\n",
    "    else:\n",
    "        raise Exception(f\"Error: {species} not founf in taxon mapper.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(csv_out, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "protspace3d",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "f967f6acf9417b335e0267c1a18089754f7cfe4000dd5b5afc6f7bca3da22610"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
